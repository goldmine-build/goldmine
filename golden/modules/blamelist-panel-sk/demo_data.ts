import { Commit } from './blamelist-panel-sk';

export const fakeNow = Date.parse('2020-03-22T00:00:00.000Z');

export const blamelist19: Commit[] = [
  {
    commit_time: 1584835000,
    hash: 'dded3c7506efc5635e60ffb7a908cbe8f1f028f1',
    author: 'Alfa (alfa@example.com)',
    message: 'Update provisioning_profile to unbreak iOS since cert refresh',
    is_cl: false,
  },
  {
    commit_time: 1584834200,
    hash: '44fc53b7f5d8390d7ae228d5fa8bac0b225ed2b9',
    author: 'Bravo (bravo@example.com)',
    message: 'Add BGR_10A2 support to Ganesh',
    is_cl: false,
  },
  {
    commit_time: 1584834100,
    hash: 'c50f4e9399cc7ae72b05eb641b50a7cd06d5af82',
    author: 'Charlie (charlie@example.com)',
    message: 'Simplify GrTextBlob::flush',
    is_cl: false,
  },
  {
    commit_time: 1584834000,
    hash: '9145f784f3261f227846e5b08dc2691a888b113c',
    author: 'Delta (deltadelta@example.com)',
    message: 'Implement D3D copySurface.',
    is_cl: false,
  },
  {
    commit_time: 1584824000,
    hash: '02b91385f58b84dc14cb6f81478b36d7fcb0d24b',
    author: 'Echo Foxtrot (ef@example.com)',
    message: 'Rename flush -> addOp',
    is_cl: false,
  },
  {
    commit_time: 1584814000,
    hash: 'a1e90aba441b0e68a07e3ab6660cd673335a76f5',
    author: 'Golf Golf Golf Golf Golf Golf Golf(golf@example.com)',
    message: 'Turn on Vulkan bots for Galaxy S20.',
    is_cl: false,
  },
  {
    commit_time: 1584804000,
    hash: 'b2b9fbc5675c46cb3af30ccd9b6e20bc57ec04c8',
    author: 'Hotel India (hotel.ind@example.com)',
    message: 'Compare all fields in SkSL::Layout::operator==',
    is_cl: false,
  },
  {
    commit_time: 1584800000,
    hash: 'f1792cde0b2b7b612cac6aa5b9e7a8ce16b9b2cb',
    author: 'Hotel India (hotel.ind@example.com)',
    message: 'Use constant swizzle syntax in GrDrawVerticesOp',
    is_cl: false,
  },
  {
    commit_time: 1584780000,
    hash: 'fc75d5c52da617b06c7566c9e389587edde3f284',
    author: 'Juliett Kilo (julia.kilo@example.com)',
    message: '[skottie] Brightness and Contrast effect',
    is_cl: false,
  },
  {
    commit_time: 1584760000,
    hash: 'e15088801e9722d9a320463554a3a21839213b48',
    author: 'Mike Lima (lima@example.com)',
    message: 'detect failed matrix update in SkDraw::drawAtlas()',
    is_cl: false,
  },
  {
    commit_time: 1584760000,
    hash: '548de7451e752ce0d6fd842c1bb4f04af4c0afdc',
    author: 'Hotel India (hotel.ind@example.com)',
    message: 'Change Marker IDs to be strings',
    is_cl: false,
  },
  {
    commit_time: 1584740000,
    hash: '7676f4ecf3ec2f544701465faca1d33af1489822',
    author: 'Zulu (zulu@example.com)',
    message: 'Remove SkFrontBufferedStream',
    is_cl: false,
  },
  {
    commit_time: 1584720000,
    hash: '4baa7326ccfbfaf66e25d91628378536d4999999',
    author: 'Papa November (pnov@example.com)',
    message: '[infra] Add POC task driver',
    is_cl: false,
  },
  {
    commit_time: 1584700000,
    hash: 'f5132a05c893a86f8bf26bcf3253985d9973fea2',
    author: 'Quebec (quebec@example.com)',
    message: "Reland \"Optimize GrTessellatePathOp's code to emit inner tri",
    is_cl: false,
  },
  {
    commit_time: 1584600005,
    hash: '3d599d16ecd8009dfb185522e87c5c1cf8a47057',
    author: 'skia-recreate-skps (skia-recreate-skps@example.com)',
    message: 'Update Go Deps',
    is_cl: false,
  },
  {
    commit_time: 1584600005,
    hash: '7a0517f8bdc0d168565a2944f0c14db06084384d',
    author: 'skia-autoroll (skia-autoroll@example.com)',
    message: 'Roll third_party/externals/angle2 3cb9c4bee9b3..4395170e6091',
    is_cl: false,
  },
  {
    commit_time: 1584600003,
    hash: 'd18100afb62401b49e0b3a5c87ed8fa006e8d3a6',
    author: 'skia-autoroll (skia-autoroll@example.com)',
    message: 'Roll ../src 9781ff27c9e9..78824aa9d99f (468 commits)',
    is_cl: false,
  },
  {
    commit_time: 1584600002,
    hash: '2ac6fa0ae796b7f050bd8e8ad60890427a4ec15e',
    author: 'skia-autoroll (skia-autoroll@example.com)',
    message: 'Roll third_party/externals/swiftshader 60aa34a990fa..2717702',
    is_cl: false,
  },
  {
    commit_time: 1584600001,
    hash: '667edf14ad72966ec36aa6cd705b98cb7d7eee28',
    author: 'skia-autoroll (skia-autoroll@example.com)',
    message: 'Roll third_party/externals/dawn 00b90ea83262..88f2ec853f80 (',
    is_cl: false,
  },
];

export const clBlamelist: Commit[] = [
  {
    commit_time: 1584600001,
    hash: '12345',
    author: 'skia-autoroll (skia-autoroll@example.com)',
    message: 'Roll third_party/externals/dawn 00b90ea83262..88f2ec853f80 (',
    is_cl: true,
  },
  {
    commit_time: 1584835000,
    hash: 'dded3c7506efc5635e60ffb7a908cbe8f1f028f1',
    author: 'Alfa (alfa@example.com)',
    message: 'Update provisioning_profile to unbreak iOS since cert refresh',
    is_cl: false,
  },
];
