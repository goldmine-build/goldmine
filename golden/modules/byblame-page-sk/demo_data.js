export const fakeNow = Date.parse('2019-11-08T00:00:00Z');

export const canvaskit = {
  data: [],
};

export const gm = {
  data: [{
    groupID: '4edb719f1bc49bae585ff270df17f08039a96b6c:252cdb782418949651cc5eb7d467c57ddff3d1c7:a1050ed2b1120613d9ae9587e3c0f4116e17337f:3f7c865936cc808af26d88bc1f5740a29cfce200:05f6a01bf9fd25be9e5fff4af5505c3945058b1d',
    nDigests: 1,
    nTests: 1,
    affectedTests: [{
      test: 'complexclip_bw_layer',
      num: 1,
      sample_digest: '457875b610908dde8bfc5f45c907eb62',
    }],
    commits: [{
      commit_time: 1573171214,
      hash: '05f6a01bf9fd25be9e5fff4af5505c3945058b1d',
      author: 'Elisa (elisa@example.com)',
      message: 'Revert Revert Reland Update commits to have subject information',
    }, {
      commit_time: 1573171074,
      hash: '3f7c865936cc808af26d88bc1f5740a29cfce200',
      author: 'Iris (iris@example.com)',
      message: 'Revert Reland Update commits to have subject information',
    }, {
      commit_time: 1573170075,
      hash: 'a1050ed2b1120613d9ae9587e3c0f4116e17337f',
      author: 'Daniel (daniel@example.com)',
      message: 'Reland Update commits to have subject information',
    }, {
      commit_time: 1573179814,
      hash: '252cdb782418949651cc5eb7d467c57ddff3d1c7',
      author: 'Elisa (elisa@example.com)',
      message: 'Revert Update commits to have subject information',
    }, {
      commit_time: 1573179564,
      hash: '4edb719f1bc49bae585ff270df17f08039a96b6c',
      author: 'Joe (joe@example.com)',
      message: 'Update commits to have subject information',
    }],
  }, {
    groupID: '4edb719f1bc49bae585ff270df17f08039a96b6c:252cdb782418949651cc5eb7d467c57ddff3d1c7',
    nDigests: 7,
    nTests: 7,
    affectedTests: [{
      test: 'aarectmodes',
      num: 1,
      sample_digest: 'c6476baec94eb6a5071606575318e4df',
    }, {
      test: 'aaxfermodes',
      num: 1,
      sample_digest: '4acfd6b3a3943cc5d75cd22e966ae6f1',
    }, {
      test: 'hairmodes',
      num: 1,
      sample_digest: 'f9e20c63b5ce3b58d9b6a90fa3e7224c',
    }, {
      test: 'imagefilters_xfermodes',
      num: 1,
      sample_digest: '47644613317040264fea6fa815af32e8',
    }, {
      test: 'lattice2',
      num: 1,
      sample_digest: '16e41798ecd59b1523322a57b49cc17f',
    }, {
      test: 'xfermodes',
      num: 1,
      sample_digest: '8fbee03f794c455c4e4842ec2736b744',
    }, {
      test: 'xfermodes3',
      num: 1,
      sample_digest: 'fed2ff29abe371fc0ec1b2c65dfb3949',
    }],
    commits: [{
      commit_time: 1573149814,
      hash: '252cdb782418949651cc5eb7d467c57ddff3d1c7',
      author: 'Elisa (elisa@example.com)',
      message: 'Reland Update commits to have subject information',
    }, {
      commit_time: 1573149564,
      hash: '4edb719f1bc49bae585ff270df17f08039a96b6c',
      author: 'Joe (joe@example.com)',
      message: 'Revert Update commits to have subject information',
    }],
  }, {
    groupID: '73a722ce97ad935f936a4c7512b6724c41e0ce4e',
    nDigests: 41,
    nTests: 1,
    affectedTests: [{
      test: 'skottie_colorize',
      num: 41,
      sample_digest: '024ce342b014c6fdb000f7c18d6d0775',
    }],
    commits: [{
      commit_time: 1572978684,
      hash: '73a722ce97ad935f936a4c7512b6724c41e0ce4e',
      author: 'Bob (bob@example.com)',
      message: 'Update commits to have subject information 4',
    }],
  }, {
    groupID: '85c3d68f2539ed7a1e71f6c9d12baaf9e6be59d8',
    nDigests: 51,
    nTests: 31,
    affectedTests: null,
    commits: [{
      commit_time: 1572899861,
      hash: '85c3d68f2539ed7a1e71f6c9d12baaf9e6be59d8',
      author: 'Alice (alice@example.com)',
      message: 'Update commits to have subject information 3',
    }],
  }, {
    groupID: '7da048b5e8f17374bcd5baf48539eaa7ebe40e5c',
    nDigests: 13,
    nTests: 3,
    affectedTests: [{
      test: 'shadow_utils',
      num: 5,
      sample_digest: '134d3c4fd609cd7f2e9cca43d78aa5d3',
    }, {
      test: 'shadow_utils_gray',
      num: 4,
      sample_digest: '292eb1e5b5860ba278ffa73efc9dd7c1',
    }, {
      test: 'shadow_utils_occl',
      num: 4,
      sample_digest: '918df0cc65d1b3ae7b8d8041afa40635',
    }],
    commits: [{
      commit_time: 1572442816,
      hash: '7da048b5e8f17374bcd5baf48539eaa7ebe40e5c',
      author: 'Frank (frank@example.com)',
      message: 'Update commits to have subject information 2',
    }],
  }, {
    groupID: '342fbc54844d0d3fc9d20e20b45115db1e33395b',
    nDigests: 1,
    nTests: 1,
    affectedTests: [{
      test: 'dftext_blob_persp',
      num: 1,
      sample_digest: '42a7bfc2d825412dfb82f6e3a12106c2',
    }],
    commits: [{
      commit_time: 1571959019,
      hash: '342fbc54844d0d3fc9d20e20b45115db1e33395b',
      author: 'Liam (liam@example.com)',
      message: 'Oh my goodness, this is an incredibly long subject line. '
        + 'How will it every fit on the page? I hope this draws ok. Some webpages '
        + 'do not work when the content is way bigger than anticipated. I hope this is '
        + 'not one of those websites',
    }],
  }],
};

export const svg = {
  data: [{
    groupID: 'd2c67f44f8c2351e60e6ee224a060e916cd44f34',
    nDigests: 12,
    nTests: 12,
    affectedTests: null,
    commits: [{
      commit_time: 1571948193,
      hash: 'd2c67f44f8c2351e60e6ee224a060e916cd44f34',
      author: 'Alice (alice@example.com)',
      message: 'Update commits to have subject information',
    }],
  }, {
    groupID: '05f6a01bf9fd25be9e5fff4af5505c3945058b1d',
    nDigests: 2,
    nTests: 1,
    affectedTests: [{
      test: 'A_large_blank_world_map_with_oceans_marked_in_blue.svg',
      num: 2,
      sample_digest: '3c62b1dd009bb18bbb84d862cbc3652c',
    }],
    commits: [{
      commit_time: 1573151114,
      hash: '05f6a01bf9fd25be9e5fff4af5505c3945058b1d',
      author: 'Elisa (elisa@example.com)',
      message: 'Update commits to have subject information',
    }],
  }, {
    groupID: 'd84dd4babb71796ee194fa1913150d86d6aa643b',
    nDigests: 2,
    nTests: 1,
    affectedTests: [{
      test: 'ynev.svg',
      num: 2,
      sample_digest: '80a7bcb4b51ad1aa876aa2144d0648bb',
    }],
    commits: [{
      commit_time: 1573048663,
      hash: 'd84dd4babb71796ee194fa1913150d86d6aa643b',
      author: 'Henry (henry@example.com)',
      message: 'Update commits to have subject information',
    }],
  }, {
    groupID: '3f7c865936cc808af26d88bc1f5740a29cfce200:05f6a01bf9fd25be9e5fff4af5505c3945058b1d',
    nDigests: 1,
    nTests: 1,
    affectedTests: [{
      test: 'A_large_blank_world_map_with_oceans_marked_in_blue.svg',
      num: 1,
      sample_digest: '02fc74ad862a6041e9ea35769a67af20',
    }],
    commits: [{
      commit_time: 1573151114,
      hash: '05f6a01bf9fd25be9e5fff4af5505c3945058b1d',
      author: 'Elisa (elisa@example.com)',
      message: 'Revert: Update commits to have subject information',
    }, {
      commit_time: 1573151074,
      hash: '3f7c865936cc808af26d88bc1f5740a29cfce200',
      author: 'Iris (iris@example.com)',
      message: 'Update commits to have subject information',
    }],
  }, {
    groupID: 'e1e197186238d8d304a39db9f94258d9584a8973',
    nDigests: 1,
    nTests: 1,
    affectedTests: [{
      test: 'gallardo.svg',
      num: 1,
      sample_digest: 'ed591512ff68fa088ec9aac3bb5d760d',
    }],
    commits: [{
      commit_time: 1572460288,
      hash: 'e1e197186238d8d304a39db9f94258d9584a8973',
      author: 'Iris (iris@example.com)',
      message: 'Update commits to have subject information',
    }],
  }],
};


export const trstatus = {
  ok: false,
  firstCommit: {
    commit_time: 1572357082,
    hash: 'ee08d523f60a04499c9023a349ef8174ab301f8f',
    author: 'Alice (alice@example.com)',
    message: 'Update commits to have subject information',
  },
  lastCommit: {
    commit_time: 1573598625,
    hash: '9501212cd0580acfed85a90c3a16b81847fde482',
    author: 'Bob (bob@example.com)',
    message: 'Revert Update commits to have subject information',
  },
  totalCommits: 256,
  filledCommits: 256,
  corpStatus: [{
    name: 'canvaskit',
    ok: false,
    minCommitHash: '',
    untriagedCount: 0,
    negativeCount: 0,
  }, {
    name: 'gm',
    ok: false,
    minCommitHash: '',
    untriagedCount: 114,
    negativeCount: 1494,
  }, {
    name: 'svg',
    ok: false,
    minCommitHash: '',
    untriagedCount: 18,
    negativeCount: 21,
  }],
};
