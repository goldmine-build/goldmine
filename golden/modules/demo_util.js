// delay wraps the value toReturn in a Promise that will resolve
// after "delay" milliseconds.
export function delay(toReturn, delay=100) {
  return new Promise((resolve) => {
    setTimeout(resolve, delay);
  }).then(() => {
    return {
      status: 200,
      body: JSON.stringify(toReturn),
      headers: {'content-type':'application/json'},
    };
  });
}