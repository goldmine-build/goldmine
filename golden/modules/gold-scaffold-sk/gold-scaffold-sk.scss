@import '~elements-sk/colors';
@import '../colors';

gold-scaffold-sk {
  app-sk {
    --header-background-color: var(--gold);
  }

  * {
    font-family:Roboto,Helvetica,Arial,'Bitstream Vera Sans',sans-serif;
  }

  header {
    h1 {
      font-weight: bold;
      font-size: 1.5em;
      padding: 0 0.25em;
    }

    .spinner-spacer {
      min-width: 40px;
    }

    spinner-sk {
      width: 24px;
      height: 24px;
      margin: 0;
      border: 8px solid var(--light-orange);
      border-left: 8px solid var(--red);
    }
  }

  nav {
    display:flex;
    flex-direction: column;
    a {
      padding: 0.5em 0.75em;
      color: var(--black);
      fill: var(--black);
      text-decoration: none;
    }
    span {
      padding: 0 0.5em;
      vertical-align: middle;
    }
  }

  login-sk {
    margin-left: auto;
    margin-right: 1em;
    display: block;
    color: white;
  }
}
