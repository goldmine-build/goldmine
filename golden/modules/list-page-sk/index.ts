import './list-page-sk';
import './list-page-sk.scss';
