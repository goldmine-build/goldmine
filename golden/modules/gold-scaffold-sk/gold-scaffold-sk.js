/**
 * @module modules/gold-scaffold-sk
 * @description <h2><code>gold-scaffold-sk</code></h2>
 *
 * Contains the title bar, side bar, and error-toast for all the Gold pages. The rest of
 * every Gold page should be a child of this element.
 *
 * Has a spinner-sk that can be activated when it hears "begin-task" events and keeps
 * spinning until it hears an equal number of "end-task" events.
 *
 * The error-toast element responds to fetch-error events and normal error-sk events.
 *
 * @attr {string} app_title - The title to show in the page banner.
 *
 * @attr {boolean} testing_offline - If we should operate entirely in offline mode.
 */
import { define } from 'elements-sk/define';
import { errorMessage } from 'elements-sk/errorMessage';
import { html } from 'lit-html';
import { ElementSk } from '../../../infra-sk/modules/ElementSk';

import '../../../infra-sk/modules/app-sk';
import '../../../infra-sk/modules/login-sk';

import 'elements-sk/error-toast-sk';
import 'elements-sk/icon/find-in-page-icon-sk';
import 'elements-sk/icon/folder-icon-sk';
import 'elements-sk/icon/help-icon-sk';
import 'elements-sk/icon/home-icon-sk';
import 'elements-sk/icon/label-icon-sk';
import 'elements-sk/icon/laptop-chromebook-icon-sk';
import 'elements-sk/icon/list-icon-sk';
import 'elements-sk/icon/search-icon-sk';
import 'elements-sk/icon/sync-problem-icon-sk';
import 'elements-sk/icon/view-day-icon-sk';
import 'elements-sk/spinner-sk';

const template = (ele) => html`
<app-sk>
  <header>
    <h1>${ele.appTitle}</h1>
    <div class=spinner-spacer>
      <spinner-sk></spinner-sk>
    </div>
    <div class=spacer></div>
    <!-- TODO(kjlubick) last commit -->
    <login-sk ?testing_offline=${ele.testingOffline}></login-sk>
  </header>

  <aside>
    <nav>
      <a href="/" tab-index=0>
        <home-icon-sk></home-icon-sk><span>Home</span>
      </a>
      <a href="/" tab-index=0>
        <view-day-icon-sk></view-day-icon-sk><span>By Blame<span>
      </a>
      <a href="/list" tab-index=0>
        <list-icon-sk></list-icon-sk><span>By Test</span>
      </a>
      <a href="/changelists" tab-index=0>
        <laptop-chromebook-icon-sk></laptop-chromebook-icon-sk><span>By ChangeList</span>
      </a>
      <a href="/search" tab-index=0>
        <search-icon-sk></search-icon-sk><span>Search</span>
      </a>
      <a href="/ignores" tab-index=0>
        <label-icon-sk></label-icon-sk><span>Ignores</span>
      </a>
      <a href="/triagelog" tab-index=0>
        <find-in-page-icon-sk></find-in-page-icon-sk><span>Triage Log</span>
      </a>
      <a href="/failures" tab-index=0>
        <sync-problem-icon-sk></sync-problem-icon-sk><span>Failures</span>
      </a>
      <a href="/help" tab-index=0>
        <help-icon-sk></help-icon-sk><span>Help</span>
      </a>
      <a href="https://github.com/google/skia-buildbot/tree/master/golden" tab-index=0>
        <folder-icon-sk></folder-icon-sk><span>Code</span>
      </a>
    </nav>
  </aside>

  <main></main>

  <footer><error-toast-sk></error-toast-sk></footer>
</app-sk>
`;

/**
 * Moves the elements from one NodeList to another NodeList.
 *
 * @param {NodeList} from - The list we are moving from.
 * @param {NodeList} to - The list we are moving to.
 */
function move(from, to) {
  Array.prototype.slice.call(from).forEach((ele) => to.appendChild(ele));
}

define('gold-scaffold-sk', class extends ElementSk {
  constructor() {
    super(template);
    this._main = null;
    this._busyTaskCount = 0;
    this._spinner = null;
  }

  connectedCallback() {
    super.connectedCallback();
    // Don't call more than once.
    if (this._main) {
      return;
    }
    this.addEventListener('begin-task', this._addBusyTask);
    this.addEventListener('end-task', this._finishedTask);
    this.addEventListener('fetch-error', this._fetchError);

    // We aren't using shadow dom so we need to manually move the children of
    // gold-scaffold-sk to be children of 'main'. We have to do this for the
    // existing elements and for all future mutations.

    // Create a temporary holding spot for elements we're moving.
    const div = document.createElement('div');
    move(this.children, div);

    // Now that we've moved all the old children out of the way we can render
    // the template.
    this._render();

    this._spinner = this.querySelector('header spinner-sk');

    // Move the old children back under main.
    this._main = this.querySelector('main');
    move(div.children, this._main);

    // Move all future children under main also.
    const observer = new MutationObserver((mutList) => {
      mutList.forEach((mut) => {
        move(mut.addedNodes, this._main);
      });
    });
    observer.observe(this, { childList: true });
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    this.removeEventListener('begin-task', this._addBusyTask);
    this.removeEventListener('end-task', this._finishedTask);
    this.removeEventListener('fetch-error', this._fetchError);
  }

  /** @prop appTitle {string} Reflects the app_title attribute for ease of use. */
  get appTitle() { return this.getAttribute('app_title'); }

  set appTitle(val) { this.setAttribute('app_title', val); }

  /** @prop {boolean} busy Indicates if there any on-going tasks (e.g. RPCs).
   *                  This also mirrors the status of the embedded spinner-sk.
   *                  Read-only. */
  get busy() { return !!this._busyTaskCount; }

  /** @prop testingOffline {boolean} Reflects the testing_offline attribute for ease of use.
   */
  get testingOffline() { return this.hasAttribute('testing_offline'); }

  set testingOffline(val) {
    if (val) {
      this.setAttribute('testing_offline', '');
    } else {
      this.removeAttribute('testing_offline');
    }
  }

  /**
   * Indicate there are some number of tasks (e.g. RPCs) the app is waiting on
   * and should be in the "busy" state, if it isn't already.
   *
   */
  _addBusyTask() {
    this._busyTaskCount++;
    if (this._spinner && this._busyTaskCount > 0) {
      this._spinner.active = true;
    }
  }

  /**
   * Removes one task from the busy count. If there are no more tasks to wait
   * for, the app will leave the "busy" state and emit the "busy-end" event.
   *
   */
  _finishedTask() {
    this._busyTaskCount--;
    if (this._busyTaskCount <= 0) {
      this._busyTaskCount = 0;
      if (this._spinner) {
        this._spinner.active = false;
      }
      this.dispatchEvent(new CustomEvent('busy-end', { bubbles: true }));
    }
  }

  /** Handles a fetch error. Expects the detail of error to contain:
   *  error: the error given to fetch.
   *  loading: A string explaining what was being fetched.
   */
  _fetchError(e) {
    const loadingWhat = e.detail.loading;
    e = e.detail.error;
    if (e.name !== 'AbortError') {
      // We can ignore AbortError since they fire anytime an AbortController was canceled.
      // Chrome and Firefox report a DOMException in this case:
      // https://developer.mozilla.org/en-US/docs/Web/API/DOMException
      console.error(e);
      errorMessage(`Unexpected error loading ${loadingWhat}: ${e.message}`,
        5000);
    }
    this._finishedTask();
  }
});
