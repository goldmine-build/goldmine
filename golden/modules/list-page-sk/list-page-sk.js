/**
 * @module module/list-page-sk
 * @description <h2><code>list-page-sk</code></h2>
 *
 * This page summarizes the outputs of various tests. It shows the amount of digests produced,
 * as well as a few options to configure what range of traces to enumerate.
 *
 * It is a top level element.
 */
import { define } from 'elements-sk/define';
import { html } from 'lit-html';
import { $$ } from 'common-sk/modules/dom';
import { jsonOrThrow } from 'common-sk/modules/jsonOrThrow';
import { stateReflector } from 'common-sk/modules/stateReflector';
import { ElementSk } from '../../../infra-sk/modules/ElementSk';
import { sendBeginTask, sendEndTask, sendFetchError } from '../common';
import { defaultCorpus } from '../settings';

import 'elements-sk/icon/group-work-icon-sk';
import 'elements-sk/icon/tune-icon-sk';
import 'elements-sk/checkbox-sk';
import '../corpus-selector-sk';
import '../query-dialog-sk';

const template = (ele) => html`
<div>
  <corpus-selector-sk .corpora=${ele._corpora}
      .selectedCorpus=${ele._currentCorpus} @corpus_selected=${ele._currentCorpusChanged}>
  </corpus-selector-sk>

  <div class=query_params>
    <button class=show_query_dialog @click=${ele._showQueryDialog}>
      <tune-icon-sk></tune-icon-sk>
    </button>
    <pre>${searchQuery(ele._currentCorpus, ele._currentQuery)}</pre>
    <checkbox-sk label="Digests at Head Only" class=head_only
             ?checked=${!ele._showAllDigests} @click=${ele._toggleAllDigests}></checkbox-sk>
    <checkbox-sk label="Disregard Ignore Rules" class=ignore_rules
             ?checked=${ele._disregardIgnoreRules} @click=${ele._toggleIgnoreRules}></checkbox-sk>
  </div>
</div>

<table>
  <!-- TODO(kjlubick) make these sortable -->
   <thead>
     <tr>
      <th>Test name</th>
      <th>Positive</th>
      <th>Negative</th>
      <th>Untriaged</th>
      <th>Total</th>
      <th>Cluster View</th>
    </tr>
  </thead>
  <tbody>
    ${ele._byTestCounts.map((row) => testRow(row, ele))}
  </tbody>
</table>

<query-dialog-sk @edit=${ele._currentQueryChanged}></query-dialog-sk>
`;

const testRow = (row, ele) => {
  const searchParams = 'unt=true&neg=true&pos=true'
    + `&source_type=${encodeURIComponent(ele._currentCorpus)}`
    + `&query=${encodeURIComponent(`name=${row.name}`)}`
    + `&head=${ele._showAllDigests ? 'false' : 'true'}`
    + `&include=${ele._disregardIgnoreRules ? 'true' : 'false'}`;

  return html`<tr>
  <td>
    <a href="/search?${searchParams}" target=_blank rel=noopener>
      ${row.name}
    </a>
  </td>
  <td class=center>${row.positive_digests}</td>
  <td class=center>${row.negative_digests}</td>
  <td class=center>${row.untriaged_digests}</td>
  <td class=center>${row.positive_digests + row.negative_digests + row.untriaged_digests}</td>
  <td class=center>
    <a href="/cluster?${searchParams}" target=_blank rel=noopener>
      <group-work-icon-sk></group-work-icon-sk>
    </a>
  </td>
</tr>`;
};

const searchQuery = (corpus, query) => {
  if (!query) {
    return `source_type=${corpus}`;
  }
  return `source_type=${corpus}, \n${query.split('&').join(',\n')}`;
};

define('list-page-sk', class extends ElementSk {
  constructor() {
    super(template);

    this._corpora = [];
    this._paramset = {};

    this._currentQuery = '';
    this._currentCorpus = '';

    this._showAllDigests = false;
    this._disregardIgnoreRules = false;

    this._stateChanged = stateReflector(
      /* getState */() => ({
        // provide empty values
        all_digests: this._showAllDigests,
        disregard_ignores: this._disregardIgnoreRules,
        corpus: this._currentCorpus,
        query: this._currentQuery,
      }), /* setState */(newState) => {
        if (!this._connected) {
          return;
        }
        // default values if not specified.
        this._showAllDigests = newState.all_digests || false;
        this._disregardIgnoreRules = newState.disregard_ignores || false;
        this._currentCorpus = newState.corpus || defaultCorpus();
        this._currentQuery = newState.query || '';
        this._fetch();
        this._render();
      },
    );

    this._byTestCounts = [];

    // Allows us to abort fetches if we fetch again.
    this._fetchController = null;
  }

  connectedCallback() {
    super.connectedCallback();
    this._render();
  }

  _currentCorpusChanged(e) {
    e.stopPropagation();
    this._currentCorpus = e.detail.corpus;
    this._stateChanged();
    this._render();
    this._fetch();
  }

  _currentQueryChanged(e) {
    e.stopPropagation();
    this._currentQuery = e.detail;
    this._stateChanged();
    this._render();
    this._fetch();
  }

  _fetch() {
    if (this._fetchController) {
      // Kill any outstanding requests
      this._fetchController.abort();
    }

    // Make a fresh abort controller for each set of fetches.
    // They cannot be re-used once aborted.
    this._fetchController = new AbortController();
    const extra = {
      signal: this._fetchController.signal,
    };

    sendBeginTask(this);
    sendBeginTask(this);

    let url = `/json/list2?corpus=${encodeURIComponent(this._currentCorpus)}`;
    if (!this._showAllDigests) {
      url += '&at_head_only=true';
    }
    if (this._disregardIgnoreRules) {
      url += '&include_ignored_traces=true';
    }
    if (this._currentQuery) {
      url += `&trace_values=${encodeURIComponent(this._currentQuery)}`;
    }
    fetch(url, extra)
      .then(jsonOrThrow)
      .then((jsonList) => {
        this._byTestCounts = jsonList;
        this._render();
        sendEndTask(this);
      })
      .catch((e) => sendFetchError(this, e, 'list'));

    // TODO(kjlubick) when the search page gets a makeover to have just the params for the given
    //   corpus show up, we should do the same here. First idea is to have a separate corpora
    //   endpoint and then make paramset take a corpus.
    fetch('/json/paramset', extra)
      .then(jsonOrThrow)
      .then((paramset) => {
        // We split the paramset into a list of corpora...
        this._corpora = (paramset.source_type || []).map((c) => ({
          name: c,
        }));
        // ...and the rest of the keys. This is to make it so the layout is
        // consistent with other pages (e.g. the search page, the by blame page, etc).
        delete paramset.source_type;
        this._paramset = paramset;
        this._render();
        sendEndTask(this);
      })
      .catch((e) => sendFetchError(this, e, 'paramset'));
  }

  _showQueryDialog() {
    $$('query-dialog-sk').open(this._paramset, this._currentQuery);
  }

  _toggleAllDigests(e) {
    e.preventDefault();
    this._showAllDigests = !this._showAllDigests;
    this._stateChanged();
    this._render();
    this._fetch();
  }

  _toggleIgnoreRules(e) {
    e.preventDefault();
    this._disregardIgnoreRules = !this._disregardIgnoreRules;
    this._stateChanged();
    this._render();
    this._fetch();
  }
});
