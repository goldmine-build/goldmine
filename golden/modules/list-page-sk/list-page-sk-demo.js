import './index';
import '../gold-scaffold-sk';
import { $$ } from 'common-sk/modules/dom';
import { delay } from '../demo_util';
import { manyParams } from '../shared_demo_data';
import { testOnlySetSettings } from '../settings';
import { sampleByTestList } from './test_data';

testOnlySetSettings({
  title: 'Testing Gold',
  defaultCorpus: 'gm',
});

const fetchMock = require('fetch-mock');

fetchMock.get('/json/paramset', delay(manyParams, 100));
fetchMock.get('glob:/json/list2*', delay(sampleByTestList, 100));

// By adding this element after all the fetches are mocked out, it should load ok.
const newList = document.createElement('list-page-sk');
const scaffold = $$('gold-scaffold-sk');
scaffold._render(); // update Title
$$('gold-scaffold-sk').appendChild(newList);
