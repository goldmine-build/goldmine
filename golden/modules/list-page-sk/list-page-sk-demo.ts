import './index';
import '../gold-scaffold-sk';
import { delay } from '../demo_util';
import { manyParams } from '../shared_demo_data';
import { testOnlySetSettings } from '../settings';
import { sampleByTestList } from './test_data';
import { exampleStatusData } from '../last-commit-sk/demo_data';
import fetchMock from 'fetch-mock';
import { ListPageSk } from './list-page-sk';
import { GoldScaffoldSk } from '../gold-scaffold-sk/gold-scaffold-sk';

testOnlySetSettings({
  title: 'Testing Gold',
  defaultCorpus: 'gm',
  baseRepoURL: 'https://github.com/flutter/flutter',
});

fetchMock.get('/json/v1/paramset', delay(manyParams, 100));
fetchMock.get('glob:/json/v1/list*', delay(sampleByTestList, 100));
fetchMock.get('/json/v1/trstatus', JSON.stringify(exampleStatusData));

// By adding these elements after all the fetches are mocked out, they should load ok.
const newScaf = new GoldScaffoldSk();
newScaf.testingOffline = true;
// Make it the first element in body.
document.body.insertBefore(newScaf, document.body.childNodes[0]);
newScaf.appendChild(new ListPageSk());
