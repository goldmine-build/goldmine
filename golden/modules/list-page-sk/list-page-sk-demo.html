<!DOCTYPE html>
<html>
<head>
  <title>list-page-sk</title>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>
<gold-scaffold-sk testing_offline>
  <!-- demo page code will insert <list-page-sk></list-page-sk> here -->
</gold-scaffold-sk>
</body>
</html>
